crates/ruma-events/src/room/message/url_preview.rs

```diff
@@ -0,0 +1,397 @@
+    #[test]
+    fn created_preview_image_to_json() {
+        let expected_result = json!({
+                  "og:image": "mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO"}
+        );
+
+        let preview =
+            PreviewImage::plain(mxc_uri!("mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO").to_owned());
+
+        assert_eq!(to_value(&preview).unwrap(), expected_result);
+
+        let encrypted_result = json!({
+            "beeper:image:encryption": {
+                "hashes" : {
+                    "sha256": "AQEBAQEBAQEBAQ",
+                },
+                "iv": "AQEBAQEBAQEBAQEB",
+                "key": {
+                    "alg": "A256CTR",
+                    "ext": true,
+                    "k": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+
+                    "key_ops": [
+                        "encrypt",
+                        "decrypt"
+                    ],
+                    "kty": "oct",
+                },
+                "v" : "v2",
+                "url": "mxc://localhost/encryptedfile",
+            },
+        });
+
+        let preview = PreviewImage::encrypted(encrypted_file());
+
+        assert_eq!(to_value(&preview).unwrap(), encrypted_result);
+    }
+
+    #[test]
+    fn parsing_regular_example() {
+        let normal_preview = json!({
+              "msgtype": "m.text",
+              "body": "https://matrix.org",
+              "m.url_previews": [
+                {
+                  "matrix:matched_url": "https://matrix.org",
+                  "matrix:image:size": 16588,
+                  "og:description": "Matrix, the open protocol for secure decentralised communications",
+                  "og:image": "mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO",
+                  "og:image:height": 400,
+                  "og:image:type": "image/jpeg",
+                  "og:image:width": 800,
+                  "og:title": "Matrix.org",
+                  "og:url": "https://matrix.org/"
+                }
+              ],
+              "m.mentions": {}
+            }
+        );
+
+        let message_with_preview: TextMessageEventContent = from_value(normal_preview).unwrap();
+        let TextMessageEventContent { url_previews, .. } = message_with_preview;
+        let previews = url_previews.expect("No url previews found");
+        assert_eq!(previews.len(), 1);
+        let UrlPreview { image, matched_url, title, url, description } = previews.first().unwrap();
+        assert_eq!(matched_url.as_ref().unwrap(), "https://matrix.org");
+        assert_eq!(title.as_ref().unwrap(), "Matrix.org");
+        assert_eq!(
+            description.as_ref().unwrap(),
+            "Matrix, the open protocol for secure decentralised communications"
+        );
+        assert_eq!(url.as_ref().unwrap(), "https://matrix.org/");
+
+        // Check the preview image parsed:
+        let PreviewImage { size, height, width, mimetype, source } = image.clone().unwrap();
```

Comment:
The clone is not necessary:

```suggestion
        let PreviewImage { size, height, width, mimetype, source } = image.as_ref().unwrap();
```